kymatio/scattering2d/backend/numpy_backend.py

```diff
@@ -0,0 +1,178 @@
+            Ideally, the last dimension should be a power of 2 to avoid errors.
+        k : int
+            integer such that x is subsampled by 2**k along the spatial variables.
+        Returns
+        -------
+        res : tensor_like
+            tensor such that its fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        batch_shape = x.shape[:-2]
+        signal_shape = x.shape[-2:]
+        
+        x = x.reshape((-1,) + signal_shape)
+        y = x.reshape(-1, k, x.shape[1] // k, k, x.shape[2] // k)
+        
+        out = np.zeros_like(y, dtype=x.dtype)
+        out = y.mean(axis=(1, 3))
+
+        return out
+
+
+class Modulus(object):
+    """
+        This class implements a modulus transform for complex numbers.
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+        Parameters
+        ---------
+        x: input complex tensor.
+        Returns
+        -------
+        output: a real tensor equal to the modulus of x.
+    """
+    def __call__(self, x):
+        norm = np.abs(x)
+        return norm
+
+
+
+
+def fft(x, direction='C2C', inverse=False):
+    """
+        Interface with torch FFT routines for 2D signals.
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x, inverse=True)
+        Parameters
+        ----------
+        input : tensor
+            complex input for the FFT
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then an error is raised.
+    """
+    if direction == 'C2R':
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if direction == 'C2R':
+        output = np.real(np.fft.ifft2(x))*x.shape[-1]*x.shape[-2]
+    elif direction == 'C2C':
+        if inverse:
+            output = np.fft.ifft2(x)*x.shape[-1]*x.shape[-2]
+        else:
+            output = np.fft.fft2(x)
+
+    return output
+
+
+
+
+def cdgmm(A, B, inplace=False):
+    """
+        Complex pointwise multiplication between (batched) tensor A and tensor B.
+        Parameters
+        ----------
+        A : tensor
+            A is a complex tensor of size (B, C, M, N, 2)
+        B : tensor
+            B is a complex tensor of size (M, N) or real tensor of (M, N)
+        inplace : boolean, optional
+            if set to True, all the operations are performed inplace
+        Returns
+        -------
+        C : tensor
+            output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
+    """
+
+    if B.ndim != 2:
+        raise RuntimeError('The dimension of the second input must be 2.')
```

Comment:
We do not have many warnings here.

Reply:
I don't understand.

Reply:
On the torch backend, CDGMM has a lot of warnings it calls. Over here we do not have too many. For example, we could check that the filter must be a 3-tensor, check that the filters are compatible for multiplication or not, check if they're of the same time.  Atleast, thats some of the stuff thats checked for on the torch backend.

Reply:
I think that makes sense. We already have the check that it's a 3-tensor though, since in NumPy, that's just a complex 2-tensor (and we have the check above).

Reply:
Yeah. Absolutely.

Reply:
Added.